packages/store/addon/-private/system/weak-cache.ts

```diff
@@ -0,0 +1,61 @@
+import { DEBUG } from '@glimmer/env';
+
+import { addSymbol, symbol } from '../ts-interfaces/utils/symbol';
+
+export default class WeakCache<K extends object, V> {
+  private _cache = new WeakMap<K, V>();
+  private _symbol?: Symbol | string;
+
+  constructor(private _fieldName: string, private _generator?: (key: K) => V, private _expectMsg?: (key: K) => string) {
+    if (DEBUG) {
+      this._symbol = symbol(_fieldName);
+    }
+  }
+
+  get(obj: K): V | undefined {
+    return this._cache.get(obj);
+  }
+
+  has(obj: object): obj is K {
+    return this._cache.has(obj as K);
+  }
+
+  delete(obj: K) {
+    return this._cache.delete(obj);
+  }
+
+  set(obj: K, value: V): void {
+    if (DEBUG && this._cache.has(obj)) {
+      throw new Error(`${obj} was already assigned a value for ${this._fieldName}`);
+    }
+    if (DEBUG) {
+      addSymbol(obj, this._symbol!, value);
+    }
+    this._cache.set(obj, value);
+  }
+
+  expect(obj: K): V {
```

Comment:
@igorT think about naming of this

Reply:
Yeah I couldn't think of a common "get and error if not present" method name. One option I've been considering is `expectGet`

Reply:
renamed to `getWithError`, open to better names but this is also entirely internal